File: Code/Python/Observer/Observer.py
class Train:
    passengers = []

    def board(self, passenger):
        passenger.enter(self)
        self.passengers += [passenger]
    
    def discharge(self, passenger):
        self.passengers.remove(passenger)
    
    def arrive(self, station):
        self.station = station
        self.announceStation()

    def announceStation(self):
        for i in range(len(self.passengers)-1, -1, -1):
            self.passengers[i].arrive(self.station)

class Passenger():
    def enter(self, train):
        self.train = train

class CountingPassenger(Passenger):
    def __init__(self, count):
        self.count = count

    def arrive(self, station):
        self.count -= 1
        if self.count == 0: self.train.discharge(self)

class CheckingPassenger(Passenger):
    def __init__(self, station):
        self.station = station

    def arrive(self, station):
        if self.station == station: self.train.discharge(self)

def trainLoop():
    stations = ["Station A", "Station B", "Station C", "Station D"]

    train = Train()
    train.station = stations[0]
    
    for passenger in [CountingPassenger(i) for i in range(1,6)]:
        train.board(passenger)

    for passenger in [CheckingPassenger(station) for station in stations]:
        train.board(passenger)

    i = 0
    while len(train.passengers) != 0:
        i = (i+1) % len(stations)
        train.arrive(stations[i])
    print(f"Final station: {train.station}")

trainLoop()

File: Code/CSharp/Observer/Observer.cs
using System;
using System.Collections.Generic;

class Train {

    public List<Passenger> passengers = new List<Passenger>();
    public string station;

    public void board(Passenger passenger) {
        passenger.enter(this);
        passengers.Add(passenger);
    }

    public void discharge(Passenger passenger) {
        passengers.Remove(passenger);
    }

    public void arrive(string station) {
        this.station = station;
        announceStation();
    }

    public void announceStation() {
        for(int i = passengers.Count - 1; i >= 0 ; i--) {
            passengers[i].arrive(station);
        }
    }
}

abstract class Passenger {
    protected Train train;
    
    public void enter(Train train) {
        this.train = train;
    }

    public abstract void arrive(string station);
}

class CountingPassenger : Passenger {
    private int count;

    public CountingPassenger(int count) {
        this.count = count;
    }
    
    public override void arrive(string station) {
        count--;
        if (count == 0) train.discharge(this);
    }
}

class CheckingPassenger : Passenger {
    public string station;
    public CheckingPassenger(string station) {
        this.station = station;
    }
    public override void arrive(string station) {
        if (this.station == station) train.discharge(this);
    }
}

class TrainLoop {
    public static void Main() {
        string[] stations = new string[]{"Station A", "Station B", "Station C", "Station D"};
        
        Train train = new Train();
        train.station = stations[0];

        for (int i = 1; i < 6; i++) {
            train.board(new CountingPassenger(i));
        }

        for (int i = 0; i < stations.Length; i++) {
            train.board(new CheckingPassenger(stations[i]));
        }

        int j = 0;
        while (train.passengers.Count != 0) {
            j = (j+1)%stations.Length;
            train.arrive(stations[j]);
        }
        Console.WriteLine("Final Station: " + train.station);
    }
}